src/StreamJsonRpc/NerdbankMessagePackFormatter.cs

```diff
@@ -0,0 +1,1425 @@
+        }
+    }
+
+    /// <inheritdoc/>
+    public object GetJsonText(JsonRpcMessage message) => message is IJsonRpcMessagePackRetention retainedMsgPack
+        ? MessagePackSerializer.ConvertToJson(retainedMsgPack.OriginalMessagePack)
+        : throw new NotSupportedException();
+
+    /// <inheritdoc/>
+    Protocol.JsonRpcRequest IJsonRpcMessageFactory.CreateRequestMessage() => new OutboundJsonRpcRequest(this);
+
+    /// <inheritdoc/>
+    Protocol.JsonRpcError IJsonRpcMessageFactory.CreateErrorMessage() => new JsonRpcError(this);
+
+    /// <inheritdoc/>
+    Protocol.JsonRpcResult IJsonRpcMessageFactory.CreateResultMessage() => new JsonRpcResult(this);
+
+    void IJsonRpcFormatterTracingCallbacks.OnSerializationComplete(JsonRpcMessage message, ReadOnlySequence<byte> encodedMessage)
+    {
+        IJsonRpcTracingCallbacks? tracingCallbacks = this.JsonRpc;
+        this.serializationToStringHelper.Activate(encodedMessage);
+        try
+        {
+            tracingCallbacks?.OnMessageSerialized(message, this.serializationToStringHelper);
+        }
+        finally
+        {
+            this.serializationToStringHelper.Deactivate();
+        }
+    }
+
+    internal static MessagePackConverter<T> GetRpcMarshalableConverter<T>()
+        where T : class
+    {
+        if (MessageFormatterRpcMarshaledContextTracker.TryGetMarshalOptionsForType(
+            typeof(T),
+            out JsonRpcProxyOptions? proxyOptions,
+            out JsonRpcTargetOptions? targetOptions,
+            out RpcMarshalableAttribute? attribute))
+        {
+            return (RpcMarshalableConverter<T>)Activator.CreateInstance(
+                typeof(RpcMarshalableConverter<>).MakeGenericType(typeof(T)),
+                proxyOptions,
+                targetOptions,
+                attribute)!;
+        }
+
+        // TODO: Improve Exception message.
+        throw new NotSupportedException($"Type '{typeof(T).FullName}' is not supported for RPC Marshaling.");
+    }
+
+    /// <summary>
+    /// Extracts a dictionary of property names and values from the specified params object.
+    /// </summary>
+    /// <param name="paramsObject">The params object.</param>
+    /// <returns>A dictionary of argument values and another of declared argument types, or <see langword="null"/> if <paramref name="paramsObject"/> is null.</returns>
+    /// <remarks>
+    /// This method supports DataContractSerializer-compliant types. This includes C# anonymous types.
+    /// </remarks>
+    [return: NotNullIfNotNull(nameof(paramsObject))]
+    private static (IReadOnlyDictionary<string, object?> ArgumentValues, IReadOnlyDictionary<string, Type> ArgumentTypes)? GetParamsObjectDictionary(object? paramsObject)
+    {
+        if (paramsObject is null)
+        {
+            return default;
+        }
+
+        // Look up the argument types dictionary if we saved it before.
+        Type paramsObjectType = paramsObject.GetType();
```

Comment:
Instead of using reflection, this method should acquire the type shape. And maybe we should just serialize it as a whole instead of breaking it up.